### DISCUSSION TITLE
`AsyncRequireAuthorizationLayer` + extractors

### DISCUSSION THREAD

**Discussion by jplatte** [tokio-rs/axum#631] [General]
It seems to me like there isn't really a way to use extractors in a closure used for `AsyncRequireAuthorizationLayer` from `tower-http`: that would require creating `RequestParts` and turning it back into `Request` later, which can fail with `axum::Error` but that can't be turned into `http::Response<_>`, which `AsyncRequireAuthorizationLayer` expects.

Did I miss something?

**Comment by davidpdrsn** [comment]
Yeah those two worlds are a little hard to bridge. Writing an extractor similar to [this](https://docs.rs/axum/0.4.2/axum/extract/fn.extractor_middleware.html#example) is probably easier.

**Reply by jplatte** [reply]
Yeah I get that that's possible but it's a bunch of boilerplate to, in an extreme example, parse the request path by hand, extract a header (handling the case where it's not there) and extract two request extensions (handling the case where they're not there). This *already* works

```rust
type Extractors =
    (Path<Uuid>, TypedHeader<Authorization<Bearer>>, Extension<Arc<AuthCache>>, Extension<PgPool>);

let request_parts = RequestParts::new(req);
let (Path(id_from_path), auth_header, auth_cache, Extension(db)) =
    Extractors::from_request(req_parts).await.map_err(IntoResponse::into_response)?;
```

it's just the final return value of the function that I can't obtain in a clean way.

Couldn't the next breaking-change release use a less generic error for `RequestParts::try_into_response`? There's really only a few cases to be handled, no?

Now that I wrote this, I realized that I can probably do this, I just need to log the `Error` and return an internal server error response! Or not..? 😅

**Reply by davidpdrsn** [reply]
> Now that I wrote this, I realized that I can probably do this, I just need to log the Error and return an internal server error response! Or not..?

That’s what I would recommend. Not being able to get the request back happens when all headers, all extensions, or the body has been consumed iirc. Logging that and returning 500 is probably fine. We could consider making the error more descriptive as well but not sure how useful it really is 🤔

**Reply by davidpdrsn** [reply]
Could change the error type to https://docs.rs/axum/0.4.2/axum/extract/rejection/enum.RequestAlreadyExtracted.html

**Reply by jplatte** [reply]
> Could change the error type to https://docs.rs/axum/0.4.2/axum/extract/rejection/enum.RequestAlreadyExtracted.html

I think this would make a lot of sense, just so it's clear that these are the only possible error conditions. 

Not that I would do anything else with it than logging or turning it into a somewhat generic json response error 😅
